core/common/src/main/java/alluxio/concurrent/ManagedBlockingUfsForwarder.java

```diff
@@ -0,0 +1,654 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.concurrent;
+
+import alluxio.AlluxioURI;
+import alluxio.SyncInfo;
+import alluxio.collections.Pair;
+import alluxio.concurrent.jsr.ForkJoinPool;
+import alluxio.security.authorization.AccessControlList;
+import alluxio.security.authorization.AclEntry;
+import alluxio.security.authorization.DefaultAccessControlList;
+import alluxio.underfs.UfsDirectoryStatus;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsMode;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.underfs.options.FileLocationOptions;
+import alluxio.underfs.options.ListOptions;
+import alluxio.underfs.options.MkdirsOptions;
+import alluxio.underfs.options.OpenOptions;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Forwarder for {@link UnderFileSystem} objects that works through with ForkJoinPool's
+ * managed blocking.
+ *
+ * If UFS calls are being done on a {@link alluxio.concurrent.jsr.ForkJoinWorkerThread}, then
```

Comment:
The idea is if there possibly is a UFS interaction, it should be wrapped in a managed blocker?

Reply:
True. Preferably an interaction that is slow.

Reply:
Based on this reasoning, please advise on all methods that you think should be or should not have been wrapped.

Reply:
Is there a downside to wrapping a call with `ManagedBlocker`? Also if a method calls another method, is it fine to have chained `ManagedBlockers`?

Reply:
Managed blocking has its overhead, so yes wrapping a fast code with `ManagedBlocker` will reduce executor's efficiency. With the current state of this code, S3 UFS has been recovered completely. So I believe there could only be methods to exclude from managed blocking.

For chain blocking, I have made an experiment by implementing `managedBlockEnter` and `managedBlockExit` methods which keep call state in TLS and don't compensate when called in a chain. I didn't see any difference between that and this usage, so continued with this instead to not increase complexity of `ForkJoinPool` by adding a custom API to it.